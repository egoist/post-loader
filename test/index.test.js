import { join } from 'path'
import webpack from 'webpack'
import MFS from 'memory-fs'
import req from 'require-from-string'
import assign from 'deep-assign'

const globalConfig = {
  output: {
    path: __dirname,
    libraryTarget: 'commonjs2',
    filename: '[name].js'
  }
}

test('no options', () => {
  const config = assign({
    entry: {
      'no-options': join(__dirname, 'fixture/no-options.md')
    },
    module: {
      rules: [{
        test: /\.md$/,
        loader: require.resolve('../src')
      }]
    }
  }, globalConfig)

  const mfs = new MFS()

  let compiler
  try {
    compiler = webpack(config)
  } catch (err) {
    return Promise.reject(new Error(err.message))
  }

  compiler.outputFileSystem = mfs

  return runCompiler(compiler)
    .then(stats => {
      expect(!stats.hasErrors() && !stats.hasWarnings())
        .toBe(true)

      const outFile = mfs.readFileSync(join(config.output.path, 'no-options.js'), 'utf8')
      const json = req(outFile)
      expect(json.title).toBe('foo')
      expect(json.content).toBe('<p>no <strong>options</strong></p>\n')
    })
})

test('custom options', () => {
  const config = assign({
    entry: {
      'custom-options': join(__dirname, 'fixture/custom-options.md')
    },
    module: {
      rules: [{
        test: /\.md$/,
        loader: require.resolve('../src'),
        options: {
          html: true
        }
      }]
    }
  }, globalConfig)

  const mfs = new MFS()
  const compiler = webpack(config)
  compiler.outputFileSystem = mfs

  return runCompiler(compiler)
    .then(stats => {
      expect(!stats.hasErrors() && !stats.hasWarnings())
        .toBe(true)

      const outFile = mfs.readFileSync(join(config.output.path, 'custom-options.js'), 'utf8')
      const json = req(outFile)
      expect(json.title).toBe('foo')
      expect(json.content).toBe('<p>custom <br/><strong>options</strong></p>\n')
    })
})

test('use plugin', () => {
  const config = assign({
    entry: {
      'use-plugin': join(__dirname, 'fixture/use-plugin.md')
    },
    module: {
      rules: [{
        test: /\.md$/,
        loader: require.resolve('../src'),
        options: {
          use: [require('markdown-it-task-lists')]
        }
      }]
    }
  }, globalConfig)

  const mfs = new MFS()
  const compiler = webpack(config)
  compiler.outputFileSystem = mfs

  return runCompiler(compiler)
    .then(stats => {
      expect(!stats.hasErrors() && !stats.hasWarnings())
        .toBe(true)

      const outFile = mfs.readFileSync(join(config.output.path, 'use-plugin.js'), 'utf8')
      expect(req(outFile).content.indexOf('task-list') !== -1).toBe(true)
    })
})

function runCompiler(compiler) {
  return new Promise((resolve, reject) => {
    compiler.run((err, stats) => {
      if (err) return reject(err)
      resolve(stats)
    })
  })
}
