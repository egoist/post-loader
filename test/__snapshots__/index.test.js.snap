// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom options 1`] = `
"module.exports =
/******/ (function(modules) { // webpackBootstrap
/******/ 	// The module cache
/******/ 	var installedModules = {};
/******/
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/
/******/ 		// Check if module is in cache
/******/ 		if(installedModules[moduleId]) {
/******/ 			return installedModules[moduleId].exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = installedModules[moduleId] = {
/******/ 			i: moduleId,
/******/ 			l: false,
/******/ 			exports: {}
/******/ 		};
/******/
/******/ 		// Execute the module function
/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);
/******/
/******/ 		// Flag the module as loaded
/******/ 		module.l = true;
/******/
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/
/******/
/******/ 	// expose the modules object (__webpack_modules__)
/******/ 	__webpack_require__.m = modules;
/******/
/******/ 	// expose the module cache
/******/ 	__webpack_require__.c = installedModules;
/******/
/******/ 	// define getter function for harmony exports
/******/ 	__webpack_require__.d = function(exports, name, getter) {
/******/ 		if(!__webpack_require__.o(exports, name)) {
/******/ 			Object.defineProperty(exports, name, {
/******/ 				configurable: false,
/******/ 				enumerable: true,
/******/ 				get: getter
/******/ 			});
/******/ 		}
/******/ 	};
/******/
/******/ 	// getDefaultExport function for compatibility with non-harmony modules
/******/ 	__webpack_require__.n = function(module) {
/******/ 		var getter = module && module.__esModule ?
/******/ 			function getDefault() { return module['default']; } :
/******/ 			function getModuleExports() { return module; };
/******/ 		__webpack_require__.d(getter, 'a', getter);
/******/ 		return getter;
/******/ 	};
/******/
/******/ 	// Object.prototype.hasOwnProperty.call
/******/ 	__webpack_require__.o = function(object, property) { return Object.prototype.hasOwnProperty.call(object, property); };
/******/
/******/ 	// __webpack_public_path__
/******/ 	__webpack_require__.p = \\"\\";
/******/
/******/ 	// Load entry module and return exports
/******/ 	return __webpack_require__(__webpack_require__.s = 0);
/******/ })
/************************************************************************/
/******/ ([
/* 0 */
/***/ (function(module, exports) {

module.exports = {\\"data\\":{\\"date\\":\\"2017-02-28T14:49:20.000Z\\",\\"title\\":\\"foo\\"},\\"content\\":\\"custom <br/>**options**\\\\n\\",\\"html\\":\\"<p>custom <br/><strong>options</strong></p>\\\\n\\"}

/***/ })
/******/ ]);"
`;

exports[`no options 1`] = `
"module.exports =
/******/ (function(modules) { // webpackBootstrap
/******/ 	// The module cache
/******/ 	var installedModules = {};
/******/
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/
/******/ 		// Check if module is in cache
/******/ 		if(installedModules[moduleId]) {
/******/ 			return installedModules[moduleId].exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = installedModules[moduleId] = {
/******/ 			i: moduleId,
/******/ 			l: false,
/******/ 			exports: {}
/******/ 		};
/******/
/******/ 		// Execute the module function
/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);
/******/
/******/ 		// Flag the module as loaded
/******/ 		module.l = true;
/******/
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/
/******/
/******/ 	// expose the modules object (__webpack_modules__)
/******/ 	__webpack_require__.m = modules;
/******/
/******/ 	// expose the module cache
/******/ 	__webpack_require__.c = installedModules;
/******/
/******/ 	// define getter function for harmony exports
/******/ 	__webpack_require__.d = function(exports, name, getter) {
/******/ 		if(!__webpack_require__.o(exports, name)) {
/******/ 			Object.defineProperty(exports, name, {
/******/ 				configurable: false,
/******/ 				enumerable: true,
/******/ 				get: getter
/******/ 			});
/******/ 		}
/******/ 	};
/******/
/******/ 	// getDefaultExport function for compatibility with non-harmony modules
/******/ 	__webpack_require__.n = function(module) {
/******/ 		var getter = module && module.__esModule ?
/******/ 			function getDefault() { return module['default']; } :
/******/ 			function getModuleExports() { return module; };
/******/ 		__webpack_require__.d(getter, 'a', getter);
/******/ 		return getter;
/******/ 	};
/******/
/******/ 	// Object.prototype.hasOwnProperty.call
/******/ 	__webpack_require__.o = function(object, property) { return Object.prototype.hasOwnProperty.call(object, property); };
/******/
/******/ 	// __webpack_public_path__
/******/ 	__webpack_require__.p = \\"\\";
/******/
/******/ 	// Load entry module and return exports
/******/ 	return __webpack_require__(__webpack_require__.s = 0);
/******/ })
/************************************************************************/
/******/ ([
/* 0 */
/***/ (function(module, exports) {

module.exports = {\\"data\\":{\\"date\\":\\"2017-02-28T14:09:59.000Z\\",\\"title\\":\\"foo\\"},\\"content\\":\\"no **options**\\\\n\\",\\"html\\":null}

/***/ })
/******/ ]);"
`;
